### PR TITLE
Add documentation for components (#65)

### LINKED ISSUE #65: Update Readme
Update readme with descriptions for spark summit, and add documentation for components

### PR DESCRIPTION
**Pull request by kevinwallimann**
Closes #65

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,106 +21,133 @@
...
-Hyperdrive is a configurable and scalable ingestion platform that allows data movement and transformation from the fast to the batch layer in a Lambda Architecture on top of Apache Spark.
-
-It is composed of two main modules: triggers and ingestors.
-
-The triggers define when ingestions must happen, the ingestor defines how.
-
-The trigger is developed in a separate repository: https://github.com/AbsaOSS/hyperdrive-trigger
+Hyperdrive is a configurable and scalable ingestion platform that allows data movement and transformation from streaming sources with exactly-once fault-tolerance semantics by using Apache Spark Structured Streaming.
 
+In Hyperdrive, each ingestion is defined by the five components reader, manager, decoder, transformer and writer. This separation allows to adapt to different streaming sources and sinks, while reusing transformations common across multiple ingestion pipelines.
```
Comment (Zejnilovic) [inline README.md]:
```suggestion
In Hyperdrive, each ingestion is defined by the five components reader, manager, decoder, transformer and writer. This separation allows adapting to different streaming sources and sinks, while reusing transformations common across multiple ingestion pipelines.
```

I am not sure about this. But this sounds "more better"

File: README.md [hunk 2]
```diff
@@ -21,106 +21,133 @@
...
-
-## Extensions
-Both, Hyperdrive trigger and ingestor are extensible.
-
-### Ingestor
+The data ingestion pipeline of Hyperdrive consists of five components: readers, managers, decoders, transformers, writers.
+- **Readers** define how to connect to sources, e.g. how to connect to Kafka in a secure cluster by providing security directives, which topic and brokers to connect to.
+- **Managers** define extra configurations for readers and writers, e.g. Kafka offset settings.
+- **Decoders** define how to convert the payload into DataFrames, e.g. decoding from binary into Avro after retrieving the schema from schema registry. 
+- **Transformers** define transformations to be applied to the decoded DataFrame, e.g. dropping columns
```
Comment (Zejnilovic) [inline README.md]:
```suggestion
- **Transformers** define transformations to be applied to the decoded DataFrame, e.g. dropping columns.
```

File: README.md [hunk 3]
```diff
@@ -21,106 +21,133 @@
...
+- **Writers** define where DataFrames should be sent after the transformations, e.g. into HDFS as Parquet files.
+
+### Built-in components
+- `KafkaStreamReader` - Connects to a secure Kafka broker.
```
Comment (Zejnilovic) [inline README.md]:
```suggestion
- `KafkaStreamReader` - connects to a secure Kafka broker.
```

File: README.md [hunk 4]
```diff
@@ -21,106 +21,133 @@
...
+- `CheckpointOffsetManager` - defines checkpoints for the stream reader and writer.
+- `ConfluentAvroKafkaStreamDecoder` - decodes the payload as Confluent Avro (through [ABRiS](https://github.com/AbsaOSS/ABRiS)), retrieving the schema from the specified Schema Registry. This decoder is capable of seamlessly handling whatever schemas the payload messages are using.
+- `ColumnSelectorStreamTransformer` - selects all columns from the decoded DataFrame.
+- `ParquetStreamWriter` - writes the DataFrame as Parquet, in **append** mode, by invoking Spark's ```processAllAvailable``` method on the stream writer.
```
Comment (Zejnilovic) [inline README.md]:
```suggestion
- `ParquetStreamWriter` - writes the DataFrame as Parquet, in **append** mode, by invoking Spark's `processAllAvailable` method on the stream writer.
```

File: README.md [hunk 5]
```diff
@@ -21,106 +21,133 @@
...
-
-Hyperdrive defines a notification topic that serves as a trigger for ingestions.
-
-The structure of the messages in the notification topic takes into account the payload topic to be offloaded into HDFS, the starting offsets, the number of messages to be processed, the destination directory, etc. 
-
-Hyperdrive supports Schema Registry connection and Confluent Avro messages (through [ABRiS](https://github.com/AbsaOSS/ABRiS)), which makes it capable of seamlessly handling whatever schemas the payload messages are using.
-
-There is also a load balancer that controls the number of active Spark ingestor jobs at a given time.
+## Usage
+Hyperdrive has to be executed with Spark. Due to Spark-Kafka integration issues, it will **only work with Spark 2.3+**.
```
Comment (Zejnilovic) [inline README.md]: and higher?

File: README.md [hunk 6]
```diff
@@ -21,106 +21,133 @@
...
+
+### How to run
+```
```
Comment (Zejnilovic) [inline README.md]:
```suggestion
```bash
```

This is the syntaxes neat feature. Can be used as well with jsno, scala, js, etc.

File: README.md [hunk 7]
```diff
@@ -21,106 +21,133 @@
...
+| :--- | :---: | :--- |
+| `reader.kafka.topic` | Yes | The name of the kafka topic to ingest data from. Equivalent to Spark property `subscribe` |
+| `reader.kafka.brokers` | Yes | List of kafka broker URLs . Equivalent to Spark property `kafka.bootstrap.servers` |
+
+Any additional properties for kafka can be added with the prefix `reader.option.`. E.g. the property `kafka.security.protocol` can be added as `reader.option.kafka.security.protocol`
+
+See e.g. the [Structured Streaming + Kafka Integration Guide](https://spark.apache.org/docs/latest/structured-streaming-kafka-integration.html) for optional kafka properties.
+
+##### CheckpointOffsetManager
+**Caution**: Currently, the `CheckpointOffsetManager` requires the property `reader.kafka.topic` to be set
```
Comment (Zejnilovic) [inline README.md]:
```suggestion
**Caution**: Currently, the `CheckpointOffsetManager` requires the property `reader.kafka.topic` to be set.
```

File: README.md [hunk 8]
```diff
@@ -21,106 +21,133 @@
...
+##### ColumnSelectorStreamTransformer
+| Property Name | Required | Description |
+| :--- | :---: | :--- |
+| `transformer.columns.to.select` | Yes |  Comma-separated list of columns to select. `*` can be used to select all columns. Only existing columns using column names may be selected (i.e. expressions cannot be constructed) |
+
+##### ParquetStreamWriter
+| Property Name | Required | Description |
+| :--- | :---: | :--- |
+| `writer.parquet.destination.directory` | Yes | Destination path of the sink. Equivalent to Spark property `path` for the `DataStreamWriter` |
+Any additional properties for the `DataStreamWriter` can be added like this: `writer.parquet.extra.conf.1=key=value`
```
Comment (Zejnilovic) [inline README.md]: Is this part of the table or not?

Comment (kevinwallimann) [reply README.md]: No

File: README.md [hunk 9]
```diff
@@ -21,106 +21,135 @@
...
+| :--- | :---: | :--- |
+| `manager.checkpoint.base.location` | Yes | Used for Spark property `checkpointLocation`. The checkpoint location is resolved to `{manager.checkpoint.base.location}/{topic}`, where `topic` is `reader.kafka.topic` |
+
+##### ConfluentAvroKafkaStreamDecoder
+The `ConfluentAvroKafkaStreamDecoder` is built on [ABRiS](https://github.com/AbsaOSS/ABRiS). More details about the configuration properties can be found there.
+
+| Property Name | Required | Description |
+| :--- | :---: | :--- |
+| `decoder.avro.schema.registry.url` | Yes | URL of Schema Registry, e.g. http://localhost:8081. Equivalent to ABRiS property `SchemaManager.PARAM_SCHEMA_REGISTRY_URL` |
+| `decoder.avro.value.schema.id=latest` | Yes | The schema id. Use `latest` or explicitly provide a number. Equivalent to ABRiS property `SchemaManager.PARAM_VALUE_SCHEMA_ID` |
```
Comment (jozefbakus) [inline README.md]: `decoder.avro.value.schema.id=latest` this should be only property name without value.

Comment (kevinwallimann) [reply README.md]: Oh yes. Thanks for spotting this one.

**Pull request merged**